File: MantidPlot/src/ProjectRecoveryGUIs/RecoveryFailureView.cpp
// Mantid Repository : https://github.com/mantidproject/mantid
//
// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
//     NScD Oak Ridge National Laboratory, European Spallation Source
//     & Institut Laue - Langevin
// SPDX - License - Identifier: GPL - 3.0 +
#include "RecoveryFailureView.h"
#include "ApplicationWindow.h"
#include "MantidKernel/UsageService.h"
#include "Script.h"
#include "ScriptingWindow.h"

RecoveryFailureView::RecoveryFailureView(QWidget *parent,
                                         ProjectRecoveryPresenter *presenter)
    : QDialog(parent), m_ui(std::make_unique<Ui::RecoveryFailure>()),
      m_presenter(presenter) {
  m_ui->setupUi(this);
  m_ui->tableWidget->horizontalHeader()->setResizeMode(QHeaderView::Stretch);
  m_ui->tableWidget->verticalHeader()->setResizeMode(QHeaderView::Stretch);
  // Make sure the ui has all the data it needs to display
  m_presenter->fillAllRows();
  // Set the table information
  addDataToTable();
  Mantid::Kernel::UsageService::Instance().registerFeatureUsage(
      "Interface", "ProjectRecoveryFailureWindow", true);
}

void RecoveryFailureView::addDataToTable() {
  // This table's size was generated for 5 which is the default but will take
  // more or less than 5, but won't look as neat
  const auto numberOfRows = m_presenter->getNumberOfCheckpoints();
  for (auto i = 0; i < numberOfRows; ++i) {
    const auto row = m_presenter->getRow(i);
    for (auto j = 0; j < row.size(); ++j) {
      m_ui->tableWidget->setItem(i, j, new QTableWidgetItem(row[j]));
    }
  }
}

void RecoveryFailureView::onClickLastCheckpoint() {
  // Recover last checkpoint
  m_presenter->recoverLast();
  Mantid::Kernel::UsageService::Instance().registerFeatureUsage(
      "Feature", "ProjectRecoveryFailureWindow->RecoverLastCheckpoint", false);
}

void RecoveryFailureView::onClickSelectedCheckpoint() {
  // Recover Selected
  QList<QTableWidgetItem *> selectedRows = m_ui->tableWidget->selectedItems();
  if (selectedRows.size() > 0) {
    const QString text = selectedRows[0]->text();
    if (text.toStdString().empty()) {
      return;
    }
    m_presenter->recoverSelectedCheckpoint(text);
  }
  Mantid::Kernel::UsageService::Instance().registerFeatureUsage(
      "Feature", "ProjectRecoveryFailureWindow->RecoverSelectedCheckpoint",
      false);
}

void RecoveryFailureView::onClickOpenSelectedInScriptWindow() {
  // Open checkpoint in script window
  QList<QTableWidgetItem *> selectedRows = m_ui->tableWidget->selectedItems();
  if (selectedRows.size() > 0) {
    const QString text = selectedRows[0]->text();
    if (text.toStdString().empty()) {
      return;
    }
    m_presenter->openSelectedInEditor(text);
  }
  Mantid::Kernel::UsageService::Instance().registerFeatureUsage(
      "Feature", "ProjectRecoveryFailureWindow->OpenSelectedInScriptWindow",
      false);
}

void RecoveryFailureView::onClickStartMantidNormally() {
  // Start save and close this, clear checkpoint that was offered for load
  m_presenter->startMantidNormally();
  Mantid::Kernel::UsageService::Instance().registerFeatureUsage(
      "Feature", "ProjectRecoveryFailureWindow->StartMantidNormally", false);
}

void RecoveryFailureView::reject() {
  // Do nothing just absorb request
  m_presenter->startMantidNormally();
  Mantid::Kernel::UsageService::Instance().registerFeatureUsage(
      "Feature", "ProjectRecoveryFailureWindow->StartMantidNormally", false);
}

void RecoveryFailureView::updateProgressBar(const int newValue,
                                            const bool err) {
  if (!err) {
    m_ui->progressBar->setValue(newValue);
  }
}

void RecoveryFailureView::setProgressBarMaximum(const int newValue) {
  m_ui->progressBar->setMaximum(newValue);
}

void RecoveryFailureView::connectProgressBar() {
  connect(&m_presenter->m_mainWindow->getScriptWindowHandle()
               ->getCurrentScriptRunner(),
          SIGNAL(currentLineChanged(int, bool)), this,
          SLOT(updateProgressBar(int, bool)));
}

void RecoveryFailureView::emitAbortScript() {
  connect(this, SIGNAL(abortProjectRecoveryScript()),
          m_presenter->m_mainWindow->getScriptWindowHandle(),
          SLOT(abortCurrent()));
  emit(abortProjectRecoveryScript());
}

void RecoveryFailureView::changeStartMantidButton(const QString &string) {
  m_ui->pushButton_3->setText(string);
}

File: qt/python/mantidqt/project/recovery/recoverygui/recoveryfailureview.py
# Mantid Repository : https://github.com/mantidproject/mantid
#
# Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
#     NScD Oak Ridge National Laboratory, European Spallation Source
#     & Institut Laue - Langevin
# SPDX - License - Identifier: GPL - 3.0 +
#  This file is part of the mantidqt package
#

from qtpy.QtWidgets import QDialog, QHeaderView, QTableWidgetItem
from qtpy.QtCore import Signal, Slot, QObject
from mantidqt.utils.qt import load_ui


class RecoveryFailureView(QDialog):
    def __init__(self, presenter):
        super(RecoveryFailureView, self).__init__()
        self.ui = load_ui(__file__, "RecoveryFailure.ui", baseinstance=self)
        self.ui.tableWidget.horizontalHeader().setSectionResizeMode(QHeaderView.Stretch)
        self.ui.tableWidget.horizontalHeader().setSectionResizeMode(QHeaderView.Stretch)
        self.presenter = presenter

        # Make sure that the presenter has all the data we need
        self.presenter.fill_all_rows()

        # Set the table data
        self._add_data_table()

    def _add_data_to_table(self):
        # This table's size was generated for 5 which is the default but will take the
        number_of_rows = self.presenter.get_number_of_checkpoints()
        for ii in range(0, number_of_rows):
            row = self.presenter.get_row(ii)
            for jj in range(0, row.size()):
                self.ui.tableWidget.setItem(ii, jj, QTableWidgetItem(row[jj]))

    def reject(self):
        self.presenter.start_mantid_normally()

    def set_progress_bar_maximum(self, new_value):
        self.ui.progressBar.setValue(new_value)

    def connect_progress_bar(self):
        QObject.connect(self.presenter.project_recovery.multi_file_interpreter.current_editor(),
                        self.presenter.project_recovery.multi_file_interpreter.current_editor().sig_progress(int),
                        self, self.update_progress_bar(int))

    def emit_abort_script(self):
        #todo: make a connection between this, with signal abort, and mainwindow's scripthandler to abort
        self.abort_project_recovery_script.emit()

    def change_start_mantid_button(self, string):
        self.ui.pushButton_3.setText(string)

    ######################################################
    #  Signals
    ######################################################

    abort_project_recovery_script = Signal()

    ######################################################
    #  Slots
    ######################################################

    @Slot(int)
    def update_progress_bar(self, new_value):
        self.ui.progressBar.setValue(new_value)

    @Slot()
    def onClickLastCheckpoint(self):
        self.presenter.recover_last()

    @Slot()
    def onClickSelectedCheckpoint(self):
        selected_rows = self.ui.tableWidget.selectedItems()
        if len(selected_rows) > 0:
            text = selected_rows[0].text()
            if text == "":
                return
            else:
                self.presenter.recover_selected_checkpoint(text)

    @Slot()
    def onClickOpenSelectedInScriptWindow(self):
        selected_rows = self.ui.tableWidget.selectedItems()
        if len(selected_rows) > 0:
            text = selected_rows[0].text()
            if text == "":
                return
            else:
                self.presenter.open_selected_in_editor(text)

    @Slot()
    def onClickStartMantidNormally(self):
        self.presenter.start_mantid_normally()